datashader/spatial/__init__.py

```diff
@@ -0,0 +1,621 @@
+            Dask dataframe containing all data from all partitions that
+            potentially intersect with the specified x_range/y_range box.
+        """
+        # Expand upper range to account for rounding
+        expanded_x_range = [x_range[0],
+                            x_range[1] + self._x_bin_width]
+
+        expanded_y_range = [y_range[0],
+                            y_range[1] + self._y_bin_width]
+
+        # Compute ranges in integer coordinates
+        query_x_range_coord = _data2coord(expanded_x_range,
+                                          self._x_range,
+                                          self._side_length)
+
+        query_y_range_coord = _data2coord(expanded_y_range,
+                                          self._y_range,
+                                          self._side_length)
+
+        # Get corresponding slice of partition grid
+        partition_query = self._partition_grid[
+            slice(*query_x_range_coord), slice(*query_y_range_coord)]
+
+        # Get unique partitions present in slice
+        query_partitions = sorted(np.unique(partition_query))
+
+        if query_partitions:
+            partition_dfs = [self._frame.get_partition(p)
+                             for p in query_partitions]
+            query_frame = dd.concat(partition_dfs)
+            return query_frame
+        else:
+            # return an empty Dask dataframe with the right shape
+            return (self._frame
+                    .get_partition(0)
+                    .map_partitions(lambda df: df.iloc[1:0]))
+
+    # Read-only properties
+    @property
+    def frame(self):
+        """
+        Dask dataframe backed by a spatially partitioned parquet file
+
+        Returns
+        -------
+        dd.DataFrame
+        """
+        return self._frame
+
+    @property
+    def hilbert_distances(self):
+        """
+        Dask series containing the Hilbert distance of each row in frame
+
+        Returns
+        -------
+        dd.Series
+        """
+        x = self.x
+        y = self._y
+        p = self._p
+        x_range = self._x_range
+        y_range = self._y_range
+        return self._frame.map_partitions(
+            _compute_distance, x=x, y=y, p=p,
+            x_range=x_range, y_range=y_range)
+
+    @property
+    def partitions(self):
+        """
+        Dask series containing the partition of each row in frame
+
+        Returns
+        -------
+        dd.Series
+        """
+        search_divisions = np.array(self.distance_divisions[:-1])
+        return self.hilbert_distances.map_partitions(
+            lambda s: pd.Series(
+                np.ones(len(s)) *
+                search_divisions.searchsorted(s.iloc[0], side='right'),
+                dtype='int64',
+                index=s.index))
+
+    @property
+    def x(self):
+        """
+        Column label in frame containing the x coordinate of each row
+        """
+        return self._x
+
+    @property
```

Comment:
Can you explain what these various properties are doing?  I assume there is a reason not to simply make them attributes?

Reply:
This might be going a bit overboard, but there are two reasons I use this approach:

 1) These properties should be read only, and if a user tries to set one they get an error.
 2) Docstrings added to the getters are visible in the notebook. E.g. `sframe.x?`

Reply:
Ok. Maybe overboard, but shouldn't hurt.